<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Available commands &mdash; expam 0.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/custom.css" type="text/css" />
    <link rel="shortcut icon" href="_static/icon.ico"/>
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
        <script>mermaid.initialize({startOnLoad:true});</script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Additional Dependencies (optional)" href="dependencies.html" />
    <link rel="prev" title="Welcome to the expam documentation!" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> expam
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#setting-the-default-database">Setting the default database</a></li>
<li class="toctree-l2"><a class="reference internal" href="#create-a-new-database">Create a new database</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#set-build-parameters">Set build parameters</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#example">Example</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#add-remove-sequences">Add/Remove sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#examples">Examples</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#printing-database-parameters">Printing database parameters</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#build-a-database">Build a database</a></li>
<li class="toctree-l2"><a class="reference internal" href="#classification">Classification</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#classify">Classify</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id1">Example</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#download-taxonomic-data">Download taxonomic data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#convert-results-to-taxonomy">Convert results to taxonomy</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="dependencies.html">Additional Dependencies (optional)</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials/index.html">Tutorials</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">expam</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Available commands</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/commands.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="available-commands">
<h1>Available commands<a class="headerlink" href="#available-commands" title="Permalink to this headline"></a></h1>
<section id="setting-the-default-database">
<h2>Setting the default database<a class="headerlink" href="#setting-the-default-database" title="Permalink to this headline"></a></h2>
<p>The EXPAM_DEFAULT_DB environment variable defines expam’s default database.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>expam default_db -db DB_NAME
<span class="go">export EXPAM_DEFAULT_DB=PATH_TO_DB</span>
</pre></div>
</div>
<p>As a shortcut,</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>expam default_db -db DB_NAME &gt;&gt; ~/.bash_profile
</pre></div>
</div>
<p>sets the environment variable each time you open bash console.</p>
</section>
<section id="create-a-new-database">
<h2>Create a new database<a class="headerlink" href="#create-a-new-database" title="Permalink to this headline"></a></h2>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>expam create -db DB_NAME
</pre></div>
</div>
<section id="set-build-parameters">
<h3>Set build parameters<a class="headerlink" href="#set-build-parameters" title="Permalink to this headline"></a></h3>
<p>Set the build parameters for some database.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>expam <span class="nb">set</span> -db DB_NAME <span class="o">[</span>args...<span class="o">]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-k">
<span id="cmdoption-kmer"></span><span class="sig-name descname"><span class="pre">-k</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;int&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--kmer</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;int&gt;</span></span><a class="headerlink" href="#cmdoption-k" title="Permalink to this definition"></a></dt>
<dd><p>K-mer size for building database.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A k-mer length of 31 is often used, and is probably good for most bacterial genomes.</p>
</div>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-n">
<span id="cmdoption-n_processes"></span><span id="cmdoption-n-processes"></span><span class="sig-name descname"><span class="pre">-n</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;int&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--n_processes</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;int&gt;</span></span><a class="headerlink" href="#cmdoption-n" title="Permalink to this definition"></a></dt>
<dd><p>Number of processes to distribute building work amongst.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Using more processes decreases the build time but increases the amount of computer
resources used during build.</p>
<p>The following short python excerpt will tell you the <strong>max</strong> number of processes.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">multiprocessing</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">multiprocessing</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
<span class="go">8</span>
</pre></div>
</div>
</div>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-p">
<span id="cmdoption-phylogeny"></span><span class="sig-name descname"><span class="pre">-p</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;file</span> <span class="pre">path&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--phylogeny</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;file</span> <span class="pre">path&gt;</span></span><a class="headerlink" href="#cmdoption-p" title="Permalink to this definition"></a></dt>
<dd><p>Path to <a href="https://en.wikipedia.org/wiki/Newick_format" target="_blank">Newick file</a> detailing tree of reference sequences.</p>
</dd></dl>

<section id="example">
<h4>Example<a class="headerlink" href="#example" title="Permalink to this headline"></a></h4>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>expam <span class="nb">set</span> -db /path/to/db -k <span class="m">31</span> -n <span class="m">12</span> -p /home/seansolari/tree.nwk
</pre></div>
</div>
</section>
</section>
<section id="add-remove-sequences">
<h3>Add/Remove sequences<a class="headerlink" href="#add-remove-sequences" title="Permalink to this headline"></a></h3>
<p>Add reference sequences to the database.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>expam add -db DB_NAME <span class="o">[</span>args...<span class="o">]</span>
<span class="gp">$ </span>expam remove -db DB_NAME <span class="o">[</span>args...<span class="o">]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-d">
<span id="cmdoption-directory"></span><span class="sig-name descname"><span class="pre">-d</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;file</span> <span class="pre">path&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--directory</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;file</span> <span class="pre">path&gt;</span></span><a class="headerlink" href="#cmdoption-d" title="Permalink to this definition"></a></dt>
<dd><p>Add sequence at <strong>file path</strong> to the database.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>File path can be a file, or a folder.</p>
<p>If a folder, expam will add all sequences within this folder.</p>
</div>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-first_n">
<span id="cmdoption-first-n"></span><span class="sig-name descname"><span class="pre">--first_n</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;int&gt;</span></span><a class="headerlink" href="#cmdoption-first_n" title="Permalink to this definition"></a></dt>
<dd><p><em>(optional)</em></p>
<p>Add first <strong>n</strong> (order same as appears in <cite>ls</cite>) sequences from directory into the database.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-group">
<span class="sig-name descname"><span class="pre">--group</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;str&gt;</span></span><a class="headerlink" href="#cmdoption-group" title="Permalink to this definition"></a></dt>
<dd><p><em>(optional)</em></p>
<p>Add sequences to particular sequence group.</p>
<p>See <a class="reference internal" href="tutorials/quickstart.html"><span class="doc">Tutorial 1</span></a> for details.</p>
</dd></dl>

<section id="examples">
<h4>Examples<a class="headerlink" href="#examples" title="Permalink to this headline"></a></h4>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>expam add -db /path/to/db -d /path/to/sequence.fna
<span class="go">Added 1 sequence from /path/to/sequence.fna!</span>

<span class="gp">$ </span>expam add -db /path/to/db -d /path/to/folder/
<span class="go">Added 19 sequences from /path/to/folder!</span>
</pre></div>
</div>
</section>
</section>
<section id="printing-database-parameters">
<h3>Printing database parameters<a class="headerlink" href="#printing-database-parameters" title="Permalink to this headline"></a></h3>
<p>Print current database configuration.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>expam print -db DB_NAME
<span class="go">&lt;&lt;&lt; expam configuration file: DB_NAME &gt;&gt;&gt;</span>

<span class="go">phylogeny       --&gt;     None</span>
<span class="go">k               --&gt;     31</span>
<span class="go">n               --&gt;     12</span>
<span class="go">sketch          --&gt;     None</span>
<span class="go">pile            --&gt;     512</span>

<span class="go">----------------</span>
<span class="go">group name: default</span>
<span class="go">        k               --&gt;     None</span>
<span class="go">        sketch          --&gt;     None</span>
<span class="go">        sequences       --&gt;     9</span>
</pre></div>
</div>
</section>
</section>
<section id="build-a-database">
<h2>Build a database<a class="headerlink" href="#build-a-database" title="Permalink to this headline"></a></h2>
<p>When all parameters have been set, a database can be built.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>expam build -db DB_NAME
</pre></div>
</div>
</section>
<section id="classification">
<h2>Classification<a class="headerlink" href="#classification" title="Permalink to this headline"></a></h2>
<section id="classify">
<h3>Classify<a class="headerlink" href="#classify" title="Permalink to this headline"></a></h3>
<p>Run metagenomic reads against a succesfully built database. See <a class="reference internal" href="tutorials/classify.html"><span class="doc">Tutorial 2</span></a> form details.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>expam run -db DB_NAME <span class="o">[</span>args...<span class="o">]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-0">
<span id="cmdoption-1"></span><span class="sig-name descname"><span class="pre">-d</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;file</span> <span class="pre">path&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--directory</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;file</span> <span class="pre">path&gt;</span></span><a class="headerlink" href="#cmdoption-0" title="Permalink to this definition"></a></dt>
<dd><p>Path containing fasta/fastq files to be classified.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Path here can either be a single file, or a folder in which case expam
will process all sequence files in this folder.</p>
</div>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-paired">
<span class="sig-name descname"><span class="pre">--paired</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-paired" title="Permalink to this definition"></a></dt>
<dd><p>To be supplied when sample files contained paired-end reads.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-name">
<span class="sig-name descname"><span class="pre">--name</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;string&gt;</span></span><a class="headerlink" href="#cmdoption-name" title="Permalink to this definition"></a></dt>
<dd><p>Name of results folder.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-taxonomy">
<span class="sig-name descname"><span class="pre">--taxonomy</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-taxonomy" title="Permalink to this definition"></a></dt>
<dd><p>Convert phylogenetic results to taxonomic results.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This requires taxonomic information for all reference sequences, see the
<a class="reference internal" href="#download-taxonomy"><span class="std std-ref">download_taxonomy</span></a> command.</p>
</div>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-cpm">
<span id="cmdoption-cutoff"></span><span class="sig-name descname"><span class="pre">--cpm</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;int&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--cutoff</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;int&gt;</span></span><a class="headerlink" href="#cmdoption-cpm" title="Permalink to this definition"></a></dt>
<dd><p>Apply cutoff to read count. <em>cpm</em> is short for counts-per-million, and takes priority if both <em>cpm</em>
and <em>cutoff</em> are supplied.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-phyla">
<span class="sig-name descname"><span class="pre">--phyla</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-phyla" title="Permalink to this definition"></a></dt>
<dd><p>Colour phylotree results by phyla.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-keep_zeros">
<span id="cmdoption-keep-zeros"></span><span class="sig-name descname"><span class="pre">--keep_zeros</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-keep_zeros" title="Permalink to this definition"></a></dt>
<dd><p>Keep nodes in output where no reads have been assigned.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-ignore_names">
<span id="cmdoption-ignore-names"></span><span class="sig-name descname"><span class="pre">--ignore_names</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-ignore_names" title="Permalink to this definition"></a></dt>
<dd><p>Don’t plot names of reference genomes in output phylotree.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-colour_list">
<span id="cmdoption-colour-list"></span><span class="sig-name descname"><span class="pre">--colour_list</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;hex</span> <span class="pre">string&gt;</span> <span class="pre">&lt;hex</span> <span class="pre">string&gt;</span> <span class="pre">...</span></span><a class="headerlink" href="#cmdoption-colour_list" title="Permalink to this definition"></a></dt>
<dd><p>List of colours to use when plotting groups in phylotree.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-2">
<span class="sig-name descname"><span class="pre">--group</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;sample</span> <span class="pre">name&gt;</span> <span class="pre">&lt;sample</span> <span class="pre">name&gt;</span> <span class="pre">...</span></span><a class="headerlink" href="#cmdoption-2" title="Permalink to this definition"></a></dt>
<dd><p>Space-separated list of sample files to be treated as a single group in phylotree.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You may also supply a hex colour code directly after the <em>–group</em> flag
to assign some colour code to this group of samples.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>expam run ... --group <span class="c1">#FF0000 sample_one sample_two</span>
</pre></div>
</div>
</div>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-alpha">
<span class="sig-name descname"><span class="pre">--alpha</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;float&gt;</span></span><a class="headerlink" href="#cmdoption-alpha" title="Permalink to this definition"></a></dt>
<dd><p>Percentage requirement for classification subtrees (see <a class="reference internal" href="tutorials/quickstart.html"><span class="doc">Tutorial 1</span></a>
and <a class="reference internal" href="tutorials/classify.html"><span class="doc">Tutorial 2</span></a>).</p>
</dd></dl>

<section id="id1">
<h4>Example<a class="headerlink" href="#id1" title="Permalink to this headline"></a></h4>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>expam run -db DB_NAME -d /path/to/paired/reads --paired --name paired_reads_analysis --taxonomy
</pre></div>
</div>
</section>
</section>
<section id="download-taxonomic-data">
<span id="download-taxonomy"></span><h3>Download taxonomic data<a class="headerlink" href="#download-taxonomic-data" title="Permalink to this headline"></a></h3>
<p>Download taxonomic metadata associated with all sequences used to build the database.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>expam download_taxonomy -db mydb
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This command can only be run after the database has been built. This is because
<strong>expam</strong> first finds NCBI accession IDs or explicit taxon IDs in the header of each
reference sequence, and uses these to search against the NCBI Entrez database.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The NCBI taxonomic ID of a reference sequence can be explicitly stated in the format</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>&gt; accessionid or metadata|taxid|TAXID|
</pre></div>
</div>
<p>For instance,</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>&gt;NZ_RJQC00000000.1|taxid|2486576|
</pre></div>
</div>
<p>If both accession ID and taxon ID are supplied, taxon ID takes precedence.</p>
</div>
</section>
<section id="convert-results-to-taxonomy">
<h3>Convert results to taxonomy<a class="headerlink" href="#convert-results-to-taxonomy" title="Permalink to this headline"></a></h3>
<p>Translate phylogenetic classification output to NCBI taxonomy.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>expam to_taxonomy --db DB_NAME
</pre></div>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="Welcome to the expam documentation!" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="dependencies.html" class="btn btn-neutral float-right" title="Additional Dependencies (optional)" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, seansolari.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>