<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Classifying metagenomic samples &mdash; expam 1.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
    <link rel="shortcut icon" href="../_static/icon.ico"/>
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
        <script>mermaid.initialize({startOnLoad:true});</script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Tree building with expam" href="treebuilding.html" />
    <link rel="prev" title="Database Build" href="build.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> expam
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../quickstart.html">Quickstart Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../dependencies.html">Additional Dependencies (optional)</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="build.html">Database Build</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Classifying metagenomic samples</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#the-classification-algorithm">The classification algorithm</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#what-do-i-do-with-splits">What do I do with splits?</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#phylogenetic-classification-results">Phylogenetic classification results</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sample-summary-files">Sample summary files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#classification-files-classified-csv">Classification files - <em>classified.csv</em></a></li>
<li class="toctree-l4"><a class="reference internal" href="#split-classification-files-split-csv">Split classification files - <em>split.csv</em></a></li>
<li class="toctree-l4"><a class="reference internal" href="#raw-read-output">Raw read output</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#taxonomic-results">Taxonomic results</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#taxonomic-sample-summaries">Taxonomic sample summaries</a></li>
<li class="toctree-l4"><a class="reference internal" href="#taxonomic-raw-output">Taxonomic raw output</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="treebuilding.html">Tree building with <strong>expam</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="graphical.html">Graphical output</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tree.html"><strong>expam</strong>’s tree module</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">expam</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="index.html">Tutorials</a> &raquo;</li>
      <li>Classifying metagenomic samples</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tutorials/classify.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="classifying-metagenomic-samples">
<h1>Classifying metagenomic samples<a class="headerlink" href="#classifying-metagenomic-samples" title="Permalink to this headline"></a></h1>
<section id="the-classification-algorithm">
<h2>The classification algorithm<a class="headerlink" href="#the-classification-algorithm" title="Permalink to this headline"></a></h2>
<ul class="simple">
<li><p>There are three types of results;</p>
<ol class="arabic simple">
<li><p>single-lineage (SL) classifications,</p></li>
<li><p><em>split</em> classifications,</p></li>
<li><p>reads that are not like anything currently in the database.</p></li>
</ol>
</li>
<li><p>Consider some metagenomic read. From this read, we can extract the corresponding k-mers.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The k-mers of a string are the set of substrings of length k. For example, the 5-mers of
<code class="code docutils literal notranslate"><span class="pre">ACGTACG</span></code> are <code class="code docutils literal notranslate"><span class="pre">ACGTA</span></code>, <code class="code docutils literal notranslate"><span class="pre">CGTAC</span></code> and <code class="code docutils literal notranslate"><span class="pre">GTACG</span></code>.</p>
</div>
<ul class="simple">
<li><p>We can map the k-mers of metagenomic reads using the <strong>expam</strong> database, where k-mers are mapped to the <em>lowest common ancestor</em> of all reference sequences containing this k-mer.</p></li>
<li><p>Due to this mapping, the k-mer distribution of any chunk of sequence from some reference genome should lie along a single lineage in the reference tree.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The k-mer distribution of a sequence corresponds to the set of points in the reference tree
that k-mers from the sequence got mapped to.</p>
</div>
<ul class="simple">
<li><p>If the k-mer distribution lies along a single lineage, this corresponds to a <em>confident</em> classification.</p>
<ul>
<li><p>In this case, the read is assigned to the lowest point of the k-mer distribution.</p></li>
<li><p><em>See Figures (a) and (b) below.</em></p></li>
</ul>
</li>
<li><p>If the k-mers diverge along multiple lineages, there are two possible explanations:</p>
<ol class="arabic simple">
<li><p>the sequence has come from a genome that is not in the reference database,</p></li>
<li><p>a sequencing error has occurred in the form of a base indel/substitution.</p></li>
</ol>
</li>
<li><p>In this case, the read is assigned to the lowest point where the lineages agree.</p></li>
<li><p><em>See Figures (c) and (d) below, and note the pressence of two lineages.</em></p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Splits can be induced in a read due to sequencing error, which may make some read of a reference genome
appear as though it does not below to the genome, as a small number of k-mers from this read will be
impacted by the incorrect base.</p>
<p>To overcome this, <strong>expam</strong> implements an <span class="math notranslate nohighlight">\(\alpha\)</span> parameter, to only consider lineages containing more than <span class="math notranslate nohighlight">\(\alpha\)</span>%
of the k-mer distribution. This should ignore those lineages in the k-mer distribution that contain
too few k-mers and are most likely due to sequencing error.</p>
</div>
<a class="reference internal image-reference" href="../_images/figure1.png"><img alt="Classification figure." class="align-center" src="../_images/figure1.png" style="width: 500px;" /></a>
<section id="what-do-i-do-with-splits">
<h3>What do I do with splits?<a class="headerlink" href="#what-do-i-do-with-splits" title="Permalink to this headline"></a></h3>
<ul>
<li><p>As we mentioned above, splits can occur either as a result of sequencing error, or due to novel sequence.</p></li>
<li><p><strong>expam</strong> implements two strategies to deal with splits as a result of sequencing error:</p>
<blockquote>
<div><ul class="simple">
<li><p>The <span class="math notranslate nohighlight">\(\alpha\)</span> parameter ignores lineages in a split distribution with low k-mer representation. <em>(See the note above.)</em></p></li>
<li><p>There are two cutoff flags that can be supplied, which filter out low abundance clades/taxa from the results. <em>These are outlined in the following note.</em></p></li>
</ul>
</div></blockquote>
</li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Sequencing errors can produce spurious classifications that manifest as low abundance clades/taxa in the summary files. These should
be filtered out before interpreting the prevalence of clades and species in your sample.</p>
<p>The <code class="code docutils literal notranslate"><span class="pre">--cutoff</span></code> flag sets a minimum count that any clade/taxa needs to reach before it is included in the classification results.
The <code class="code docutils literal notranslate"><span class="pre">--cpm</span></code> flag sets the same cutoff, but as a rate of <strong>count required per million reads in the sample</strong>, as opposed to a flat cutoff number.</p>
<p>When both are supplied, the highest of either cutoff is taken. <em>By default</em>, <strong>expam</strong> <em>requires each node to have at least
100 counts per million input reads</em>.</p>
</div>
<ul class="simple">
<li><p>With both these mechanisms in place, we can be more confident that high split counts in a particular region of the phylogeny is suggestive of novel sequence in the biological sample.</p></li>
<li><p>The algorithm for classifying splits takes a conservative approach - <strong>those that are interested only in a general profile can feel comfortable simply adding classification and split counts together to produce an overall profile.</strong></p></li>
<li><p><em>Splits</em> can also be used as a marker for genome discovery - samples reported with a high split counts are potential targets for culturing novel isolates.</p></li>
</ul>
</section>
</section>
<section id="phylogenetic-classification-results">
<h2>Phylogenetic classification results<a class="headerlink" href="#phylogenetic-classification-results" title="Permalink to this headline"></a></h2>
<ul>
<li><p>Say we have just run a sample <code class="code docutils literal notranslate"><span class="pre">sample_one.fq</span></code> against the database, and the classification results are in a folder <code class="code docutils literal notranslate"><span class="pre">./sample_one</span></code>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>expam classify -db my_database -d /path/to/sample_one.fq --out sample_one
</pre></div>
</div>
</li>
<li><p>In <code class="code docutils literal notranslate"><span class="pre">./sample_one</span></code>, there will be a <code class="code docutils literal notranslate"><span class="pre">phy</span></code> subdirectory containing three files:</p>
<ul class="simple">
<li><p><code class="code docutils literal notranslate"><span class="pre">./sample_one/phy/sample_one.csv</span></code> - sample summary file.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">./sample_one/phy/classified.csv</span></code> - complete classifications.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">./sample_one/phy/split.csv</span></code> - split classifications.</p></li>
<li><p>Within <code class="code docutils literal notranslate"><span class="pre">./sample_one/phy</span></code>, there will be a <code class="code docutils literal notranslate"><span class="pre">raw</span></code> subdirectory containing the output for each read.</p></li>
</ul>
</li>
</ul>
<section id="sample-summary-files">
<h3>Sample summary files<a class="headerlink" href="#sample-summary-files" title="Permalink to this headline"></a></h3>
<ul>
<li><p>Each input sample file gets a corresponding sample summary file.</p></li>
<li><p>Comma-separated file of all results for the sample, both complete and split.</p></li>
<li><p>There are seven columns:</p>
<ol class="arabic simple">
<li><p><strong>Node</strong> - classification point in tree.</p></li>
</ol>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If a node points to some location in the reference tree, it will start with a ‘p’.</p>
<p>This point can be used as input to the <a class="reference internal" href="../tree.html"><span class="doc">programmatic tree interface</span></a> for further analysis.</p>
</div>
<ol class="arabic simple" start="2">
<li><p><strong>Percent classified (cumulative)</strong> - the percentage of all reads classified at or below this node.</p></li>
<li><p><strong>Total classified (cumulative)</strong> - the raw number of reads classified at or below this node.</p></li>
<li><p><strong>Total classified (raw)</strong> - the total number of reads classified at precisely this point.</p></li>
<li><p><strong>Percent split (cumulative)</strong> - the percentage of all reads classified as a split at or below this point.</p></li>
<li><p><strong>Total split (cumulative)</strong> - the total number of reads classified as a split at or below this point.</p></li>
<li><p><strong>Total split (raw)</strong> - the number of reads classified as a split precisely at this node.</p></li>
</ol>
</li>
</ul>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The first row of this file contols those that are unclassified - neither classified nor split.</p>
</div>
<section id="example">
<h4>Example<a class="headerlink" href="#example" title="Permalink to this headline"></a></h4>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">=================================== =================================== ============================== ======================= ============================== ========================= ==================</span>
<span class="go">Node                                Cumulative Classified Percentage    Cumulative Classified Count    Raw Classified Count    Cumulative Split Percentage    Cumulative Split Count    Raw Split Count</span>
<span class="go">=================================== =================================== ============================== ======================= ============================== ========================= ==================</span>
<span class="go">unclassified                        0.0%                                0                              0                       0.0%                           0                         0</span>
<span class="go">p1                                  100.0%                              1000                           3                       0.0%                           0                         0</span>
<span class="go">p2                                  99.7%                               997                            232                     0.0%                           0                         0</span>
<span class="go">p5                                  76.5%                               765                            0                       0.0%                           0                         0</span>
<span class="go">GCF_000005845.2_ASM584v2_genomic    76.5%                               765                            765                     0.0%                           0                         0</span>
<span class="go">=================================== =================================== ============================== ======================= ============================== ========================= ==================</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>By default, only nodes with counts (above the cutoff) will be included in these summaries. To include all nodes,
add the <code class="code docutils literal notranslate"><span class="pre">--keep-zeros</span></code> flag at classification.</p>
</div>
</section>
</section>
<section id="classification-files-classified-csv">
<h3>Classification files - <em>classified.csv</em><a class="headerlink" href="#classification-files-classified-csv" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Comma-separated matrix - cells contain number of reads classified to specific node (row) within any given sample (column).</p></li>
<li><p>This enables phylogenetic comparison of samples.</p></li>
<li><p>These classifications correspond to those reads whose k-mer distribution lies on a single lineage (high quality).</p></li>
</ul>
<section id="id1">
<h4>Example<a class="headerlink" href="#id1" title="Permalink to this headline"></a></h4>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">=================================== ========================================</span>
<span class="go">Node                                Sample One</span>
<span class="go">=================================== ========================================</span>
<span class="go">unclassified                        0</span>
<span class="go">p1                                  3</span>
<span class="go">p2                                  232</span>
<span class="go">GCF_000005845.2_ASM584v2_genomic    765</span>
<span class="go">=================================== ========================================</span>
</pre></div>
</div>
</section>
</section>
<section id="split-classification-files-split-csv">
<h3>Split classification files - <em>split.csv</em><a class="headerlink" href="#split-classification-files-split-csv" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Comma-separated matrix with same interpretation as <em>classified.csv</em>, only these results are those classifications whose lineage was split.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The rows and columns of <em>classified.csv</em> and <em>split.csv</em> will always line up with eachother.</p>
<p>This is for convenience - those who simply want an overall phylogenetic profile can add these two matrices together without
needing to pre-process and align the corresponding rows and columns.</p>
</div>
</section>
<section id="raw-read-output">
<h3>Raw read output<a class="headerlink" href="#raw-read-output" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>Contains the read-wise output for each input sample file.</p></li>
<li><p>Each file is tab-delimited, with five columns:</p>
<ol class="arabic simple">
<li><p><strong>Classification code</strong> - one of C (classified), S (split) or U (unclassified).</p></li>
<li><p><strong>Read ID</strong> - unique identifier for each read, taken from the header line of each sequence.</p></li>
<li><p><strong>Node</strong> - the phylogenetic node where each read is classified.</p></li>
<li><p><strong>Read length</strong> - length of the read string.</p></li>
<li><p><strong>Classification breakdown</strong> - this formatted string is a space-delimited summary of where the kmers of this read belonged to. For instance, the summary <code class="code docutils literal notranslate"><span class="pre">p4:5</span> <span class="pre">p8:16</span> <span class="pre">p4:198</span></code> means that 5 kmers were assigned to node <code class="code docutils literal notranslate"><span class="pre">p4</span></code>, 16 kmers were then assigned to node <code class="code docutils literal notranslate"><span class="pre">p8</span></code>, and finally another 198 kmers were again assigned to <code class="code docutils literal notranslate"><span class="pre">p4</span></code>. These results are reported in order, reading the sequence from left to right.</p></li>
</ol>
</li>
</ul>
</section>
</section>
<section id="taxonomic-results">
<h2>Taxonomic results<a class="headerlink" href="#taxonomic-results" title="Permalink to this headline"></a></h2>
<ul class="simple">
<li><p>Provided you have run the <code class="code docutils literal notranslate"><span class="pre">download_taxonomy</span></code> command (see <a class="reference internal" href="../commands.html#download-taxonomy"><span class="std std-ref">section in Commands documentation</span></a>), you can convert the above phylogenetic results into the taxonomic setting.</p></li>
<li><p>The following two commands accomplish this task equivalently:</p></li>
</ul>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>expam classify -d /path/to/reads --out example --taxonomy
</pre></div>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>expam classify -d /path/to/reads --out example
<span class="gp">$ </span>expam to_taxonomy --out example
</pre></div>
</div>
<ul class="simple">
<li><p>Where before the results directory contained only a <code class="code docutils literal notranslate"><span class="pre">phy</span></code> subdirectory, <strong>expam</strong> will now also create a <code class="code docutils literal notranslate"><span class="pre">tax</span></code> folder, which will now be populated with the corresponding taxonomic output.</p></li>
</ul>
<section id="taxonomic-sample-summaries">
<h3>Taxonomic sample summaries<a class="headerlink" href="#taxonomic-sample-summaries" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>For each sample input file, <strong>expam</strong> will translate a corresponding taxonomic sample summary.</p></li>
<li><p>These are comma-separated matrices with nine columns:</p>
<ol class="arabic simple">
<li><p><strong>Taxon ID</strong> - NCBI taxon id.</p></li>
<li><p><strong>Percent classified (cumulative)</strong> - total percentage of reads in this sample classified at or below this taxon id.</p></li>
<li><p><strong>Total classified (cumulative)</strong> - total number of reads classified at or below this taxon id.</p></li>
<li><p><strong>Classified (raw)</strong> - number of reads classified directly to this taxon id.</p></li>
<li><p><strong>Percent split (cumulative)</strong> - total percentage of reads classified as a split, at or below this taxon id.</p></li>
<li><p><strong>Total split (cumulative)</strong> - total number of reads classified as a split, at or below this taxon id.</p></li>
<li><p><strong>Split (raw)</strong> - number of reads classified as a split directly at this taxon id.</p></li>
<li><p><strong>Rank</strong> - taxonomic rank associated with the taxon id.</p></li>
<li><p><strong>Scientific name (lineage)</strong> - (space-separated) taxonomic lineage associated with this taxon id.</p></li>
</ol>
</li>
</ul>
<section id="id2">
<h4>Example<a class="headerlink" href="#id2" title="Permalink to this headline"></a></h4>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">=============== =================================== ============================== ======================= ============================== ========================= ================== =============== =================================================================</span>
<span class="go">Node            Cumulative Classified Percentage    Cumulative Classified Count    Raw Classified Count    Cumulative Split Percentage    Cumulative Split Count    Raw Split Count    Rank            Scientific Name</span>
<span class="go">=============== =================================== ============================== ======================= ============================== ========================= ================== =============== =================================================================</span>
<span class="go">unclassified    0.0%                                0                              0                       0.0%                           0                         0                  0               0</span>
<span class="go">1               100.0%                              1000                           0                       0.0%                           0                         0                  root</span>
<span class="go">131567          100.0%                              1000                           0                       0.0%                           0                         0                  top             cellular organisms</span>
<span class="go">2               100.0%                              1000                           235                     0.0%                           0                         0                  superkingdom    cellular organisms Bacteria</span>
<span class="go">1224            76.5%                               765                            0                       0.0%                           0                         0                  phylum          cellular organisms Bacteria Proteobacteria</span>
<span class="go">1236            76.5%                               765                            0                       0.0%                           0                         0                  class           cellular organisms Bacteria Proteobacteria Gammaproteobacteria</span>
<span class="go">=============== =================================== ============================== ======================= ============================== ========================= ================== =============== =================================================================</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><strong>expam</strong> only supplies taxonomic versions for sample summary files, it does not create any
taxonomic version of the <code class="code docutils literal notranslate"><span class="pre">classified.csv`</span></code> or <code class="code docutils literal notranslate"><span class="pre">splits.csv</span></code> as in the phylogenetic case.</p>
</div>
</section>
</section>
<section id="taxonomic-raw-output">
<h3>Taxonomic raw output<a class="headerlink" href="#taxonomic-raw-output" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p><strong>expam</strong> also translates raw classification outputs for each read into the taxonomic setting.</p></li>
<li><p>This is located in <code class="code docutils literal notranslate"><span class="pre">../run_name/tax/raw</span></code>, again with one summary file per sample.</p></li>
<li><p>There are four tab-delimited columns:</p>
<ol class="arabic simple">
<li><p><strong>Classification Code</strong></p></li>
<li><p><strong>Read ID</strong> - unique identifier for each read, taken from header lines of the sequence.</p></li>
<li><p><strong>Taxon ID</strong> - NCBI taxon id that this read was assigned to.</p></li>
<li><p><strong>Read length</strong> - length of the read string.</p></li>
</ol>
</li>
</ul>
<section id="id3">
<h4>Example<a class="headerlink" href="#id3" title="Permalink to this headline"></a></h4>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">C</span>       <span class="n">R4825323246286034638</span>    <span class="mi">543</span>     <span class="mi">302</span>
<span class="n">C</span>       <span class="n">R4280015672552393909</span>    <span class="mi">511145</span>  <span class="mi">302</span>
<span class="n">S</span>       <span class="n">R5925738157954038177</span>    <span class="mi">511145</span>  <span class="mi">302</span>
<span class="n">C</span>       <span class="n">R3237657389899545456</span>    <span class="mi">511145</span>  <span class="mi">302</span>
<span class="n">C</span>       <span class="n">R6111671585932593081</span>    <span class="mi">511145</span>  <span class="mi">302</span>
<span class="n">C</span>       <span class="n">R4574482278193488645</span>    <span class="mi">511145</span>  <span class="mi">302</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Note the lack of ‘p’ at the start of values in the third column - these refer to NCBI taxonomic IDs,
not points in the reference tree.</p>
</div>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="build.html" class="btn btn-neutral float-left" title="Database Build" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="treebuilding.html" class="btn btn-neutral float-right" title="Tree building with expam" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, seansolari.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>