## Integrating metagenomics with host transcriptomics

### Required modules

``` R
library(readr)
library(reshape)
library(tidyverse)
```

``` R
source("scripts/preprocessing/preprocessing.R")
source("scripts/taxonomy/taxonomy.R")
```

Functions provided by `preprocessing.R`

**clr** - computes a centered log-ratio transformation of the metagenomics count data.

Functions provided by `taxonomy.R`

**load_taxonomy_files** - load the taxonomic assignment files computed within an *expam* database, with the `expam download_taxonomy [args ...]` command.

**make_lineage_table** - computes a table of taxonomic lineages, where columns are prominent taxonomic ranks (Phylum, Class, Order, Family, Genus, Species), and rows are different taxonomic IDs. 


### Input files

The `clade_taxonomy_file` is a table of taxonomic assignments for each phylogenetic clade generated by the `scripts/taxonomy/mapper.py` script.

``` R
metadata_file <- "data/metadata.csv"
metagenomics_file <- "data/expam_output/clean/cumulative_cpm5.csv"
transcriptomics_file <- "data/rnaseq/DGEHallmarkScore99ed.txt"
db_path <- "data/expam_database/pibd"
clade_taxonomy_file <- "data/expam_database/taxonomy.csv"
```

### Load datasets

``` R
meta <- readr::read_csv(metadata_file)

X <- readr::read_csv(metagenomics_file) %>%
    rename(clade = `...1`)
Y <- readr::read_tsv(transcriptomics_file)
```

Align samples in `X` and `Y` dataframes

``` R
X_samples <- colnames(X[,-1])
Y_samples <- colnames(Y[,-1])
X_AND_Y_samples <- base::intersect(X_samples, Y_samples)

X_mask <- c(1, match(X_AND_Y_samples, X_samples)+1)
Y_mask <- c(1, match(X_AND_Y_samples, Y_samples)+1)

X_clean <- X[,X_mask]
Y_clean <- Y[,Y_mask]
```

### Load taxonomic data

``` R
# Mutate specific rows of a DataFrame that match a condition
# See https://stackoverflow.com/q/34096162
mutate_cond <- function(.data, condition, ..., envir = parent.frame()) {
    condition <- eval(substitute(condition), .data, envir)
    .data[condition, ] <- .data[condition, ] %>% mutate(...)
    .data
}

tax <- load_taxonomy_files(db_path)
taxid_data <- make_lineage_table(taxonomy, propagate = T)
```

Map taxonomic assignments for each clade using  `clade_taxonomy_file`.

``` R
tax$clades <- read_csv(clade_taxonomy_file, col_names = c("clade", "taxid")) %>%
    mutate_cond(!is.na(as.numeric(clade)),
                clade = paste0("p", clade))
```

### Species-level clades

To reduce the dimensionality of the dataset, we restrict the analysis to species-level clades. 

**Note that this is not grouping the input counts into species** - the metagenomics feature variables still correspond to phylogenetic clades in the reference tree, we simply restrict our focus to those clades that correspond to species-level taxonomic relationships (as dictated by *expam*'s taxonomic conversion algorithm).

``` R
# get list of clades corresponding to taxonomic species
clade_ranks <- colmap(tax$clades$taxid, tax$taxa_rank$taxid, tax$taxa_rank$rank)
clade_rank_inds <- match(clade_ranks, all_ranks)
species_ind <- match("species", all_ranks)
mask <- clade_rank_inds == species_ind
species_nodes <- tax$clades$clade[mask]

# subset data using this list
X_species <- X_clean %>%
    filter(clade %in% species_nodes)
```

Put data into format for integration

``` R
# Convert to appropriately aligned matrices
MX_sp <- as.matrix(X_species[,-1])
rownames(MX_sp) <- X_species %>% pull(clade)

MY <- as.matrix(Y_clean[,-1])
rownames(MY) <- Y_clean %>% pull(Set)
```

Apply Centered Log-Ratio transformation on metagenomic counts

``` R
MX_sp_clr <- clr(MX_sp)

X_mgm <- t(MX_sp_clr)
Y_rna <- t(MY)
```

### Sparse Partial Least Squares Integration

Integration performed using [mixOmics](http://mixomics.org/) package.

``` R
library(mixOmics)
```

Initial PCA to get an idea of the number of components required

``` R
pca.mgm <- pca(X_mgm, ncomp = 10, center = F, scale = F)
pca.rna <- pca(Y_rna, ncomp = 10, center = T, scale = T)

plot(pca.mgm)
plot(pca.rna)
```

Run sPLS integration 

``` R
pibd.spls <- spls(
    X = X_mgm,
    Y = Y_rna,
    ncomp = 5,
    keepX = c(15, 15, 15, 15, 15),
    keepY = c(5, 5, 5, 5, 5),
    mode = "regression"
)
pibd.spls.cim <- cim(pibd.spls, comp = 1:5, margins = c(19, 10), row.cex = 0.2, col.cex = 0.8)
```

### Plot resulting heatmap of associations

``` R
htmap <- pibd.spls.cim$mat
colnames(htmap)
rownames(htmap)

# colour generation
col_generator <- rev(RColorBrewer::brewer.pal(10, "Spectral"))

# melt dataframe for plotting
plot_df <- melt(htmap)
colnames(plot_df) <- c("clade", "geneset", "score")

# assign genus and species labels
htmap_taxids <- colmap(plot_df$clade, tax$clades$clade, tax$clades$taxid)
plot_df$species <- colmap(htmap_taxids, tax$taxa_rank$taxid, tax$taxa_rank$taxa)
plot_df$genus <- sapply(strsplit(plot_df$species, " "), function(x) x[1])

# make ggplot heatmap
ht <- ggplot(plot_df, aes(x = geneset, y = clade, fill = score)) +
    labs(x = element_blank(), y = element_blank(), title = element_blank()) +
    theme(
        axis.text.x = element_text(angle = 65, hjust = 1),
        axis.ticks = element_blank(),
        panel.background = element_blank(),
        panel.spacing = unit(3, "pt"),
        strip.background = element_blank(),
        strip.placement = "inside",
        strip.text.y = element_text(
            colour = "black",
            face = "italic",
            size = 6,
            hjust = 1),
        strip.text.y.left = element_text(angle = 0)) +
    geom_tile(color = NA, height = 1.0, linewidth = 0.0) +
    facet_grid(genus ~ ., space = "free", scales = "free", switch = "y") +
    scale_fill_gradientn(colors = col_generator) +
    scale_y_discrete(labels = element_blank())
```

